### PR TITLE
Add Job Gauge event

### PR DESCRIPTION
**Pull request by valarnin**
This PR adds a Job Gauge Changed event. This is part of upstreaming cactbot code. I believe that after this is upstreamed, all of cactbot's memory reading code can be replaced with OverlayPlugin information.

For reference, the reason we need an event despite there already being an `FFXIV_ACT_Plugin` event for network gauge information, is because that line only contains partial data (basically no timer-related information).

Example below. Note that `BaseObject` is a debug property and will be excluded with `#ifndef` once I'm done testing.

`RawData` maybe could be converted to `ulong`, or properly converted to a hex string. I'm not sure yet how to handle it.

```json
[
  {
    "Job": 24,
    "Data": {
      "LilyTimer": 12645,
      "Lily": 0,
      "BloodLily": 0
    },
    "RawData": "AABlMQAAAAA=",
    "BaseObject": {
      "EmptyGauge": {},
      "WhiteMage": {
        "LilyTimer": 12645,
        "Lily": 0,
        "BloodLily": 0
      },
      "Scholar": {
        "Aetherflow": 0,
        "FairyGauge": 0,
        "SeraphTimer": 12645,
        "DismissedFairy": 0
      },
      "Astrologian": {
        "Timer": 0,
        "Card": 0,
        "Seals": 0,
        "CurrentCard": 0,
        "CurrentSeals": [
          0,
          0,
          0
        ]
      },
      "Sage": {
        "AddersgallTimer": 0,
        "Addersgall": 101,
        "Addersting": 49,
        "Eukrasia": 0,
        "EukrasiaActive": false
      },
      "Bard": {
        "SongTimer": 0,
        "Repertoire": 0,
        "SoulVoice": 0,
        "SongFlags": 0
      },
      "Machinist": {
        "OverheatTimeRemaining": 0,
        "SummonTimeRemaining": 12645,
        "Heat": 0,
        "Battery": 0,
        "LastSummonBatteryPower": 0,
        "TimerActive": 0
      },
      "Dancer": {
        "Feathers": 0,
        "Esprit": 0,
        "DanceSteps": {
          "FixedElementField": 101
        },
        "StepIndex": 0,
        "CurrentStep": 101
      },
      "BlackMage": {
        "EnochianTimer": 0,
        "ElementTimeRemaining": 12645,
        "ElementStance": 0,
        "UmbralHearts": 0,
        "PolyglotStacks": 0,
        "EnochianFlags": 0,
        "UmbralStacks": 0,
        "AstralStacks": 0,
        "EnochianActive": false,
        "ParadoxActive": false
      },
      "Summoner": {
        "SummonTimer": 0,
        "AttunementTimer": 12645,
        "ReturnSummon": 0,
        "ReturnSummonGlam": 0,
        "Attunement": 0,
        "AetherFlags": 0
      },
      "RedMage": {
        "WhiteMana": 0,
        "BlackMana": 0,
        "ManaStacks": 101
      },
      "Monk": {
        "Chakra": 0,
        "BeastChakra1": 0,
        "BeastChakra2": 101,
        "BeastChakra3": 49,
        "Nadi": 0,
        "BlitzTimeRemaining": 0,
        "BeastChakra": [
          0,
          101,
          49
        ]
      },
      "Dragoon": {
        "LotdTimer": 0,
        "LotdState": 101,
        "EyeCount": 49,
        "FirstmindsFocusCount": 0
      },
      "Ninja": {
        "HutonTimer": 0,
        "Ninki": 101,
        "HutonManualCasts": 49
      },
      "Samurai": {
        "Kaeshi": 101,
        "Kenki": 49,
        "MeditationStacks": 0,
        "SenFlags": 0
      },
      "Reaper": {
        "Soul": 0,
        "Shroud": 0,
        "EnshroudedTimeRemaining": 12645,
        "LemureShroud": 0,
        "VoidShroud": 0
      },
      "DarkKnight": {
        "Blood": 0,
        "DarksideTimer": 12645,
        "DarkArtsState": 0,
        "ShadowTimer": 0
      },
      "Paladin": {
        "OathGauge": 0
      },
      "Warrior": {
        "BeastGauge": 0
      },
      "Gunbreaker": {
        "Ammo": 0,
        "MaxTimerDuration": 12645,
        "AmmoComboStep": 0
      },
      "ClassJobID": 24,
      "GetRawGaugeData": "AABlMQAAAAA="
    },
    "type": "JobGaugeChanged"
  }
]
```

### CODE REVIEW

**Comment by xpdota** [comment]
I suppose RawData could also be formatted as an array of bytes. Looks good to me.

**Comment by valarnin** [comment]
wexxlee, please feel free to offer additional feedback on e.g. the exact return value here or the polling rate, as the intention is to eventually replace cactbot's remaining memory reading code with this event so it needs to meet all of the requirements from that angle.

**Comment by valarnin** [comment]
Force push/rebase to make sure there are no build warnings on this now that the other PR cleaning them up has been merged.

**Comment by wexxlee** [comment]
> wexxlee, please feel free to offer additional feedback on e.g. the exact return value here or the polling rate, as the intention is to eventually replace cactbot's remaining memory reading code with this event so it needs to meet all of the requirements from that angle.

This looks great to me.  I think 50ms polling is good -- the jobs module really shouldn't need anything more frequent than that, so as long as that frequency isn't causing excessive lag, then I think it's good.  In that vein, would it be worthwhile to add a counter on `Thread.Sleep()` so that if it reaches some X threshold over Y time (e.g. persistent lag), the polling rate is decreased and something gets written to the console, so there's something there for troubleshooting?  That said, iirc, we don't do that with any of the other memory polling stuff, so maybe it's not worth creating an outlier.  Feel free to ignore, just thinking out loud.

When we start stripping down cactbot, we can always revisit anything here if it looks like anything is missing, but I think this is good to land.

**Review by wexxlee — APPROVED** [review]

**Comment by xpdota** [comment]
Instead of rapid polling, could it trigger any time it gets a job gauge line + have a slow polling as a backup?

**Comment by valarnin** [comment]
> Instead of rapid polling, could it trigger any time it gets a job gauge line + have a slow polling as a backup?

cactbot currently polls this data every 16ms, see:
https://github.com/OverlayPlugin/cactbot/blob/72aa4ba56523bd496ab1b3f51416bed0db47a514/plugin/CactbotEventSource/CactbotEventSource.cs#L19

All of the actual overlay-side logic relies on this event firing every 16ms (or rather, there is no actual timer logic in the overlay itself, it just refreshes with the most recent event data whenever it receives it). This would require significant rewriting of logic for the jobs overlay.

I'm not opposed to this or anything, I just wanted to make sure you understood the scope that such a change would involve. I agree that it would be better long-term to only poll memory on a log line, but that's a big difference compared to the functionality this is meant to replicate.

**Comment by xpdota** [comment]
> > Instead of rapid polling, could it trigger any time it gets a job gauge line + have a slow polling as a backup?
> 
> cactbot currently polls this data every 16ms, see: https://github.com/OverlayPlugin/cactbot/blob/72aa4ba56523bd496ab1b3f51416bed0db47a514/plugin/CactbotEventSource/CactbotEventSource.cs#L19
> 
> All of the actual overlay-side logic relies on this event firing every 16ms (or rather, there is no actual timer logic in the overlay itself, it just refreshes with the most recent event data whenever it receives it). This would require significant rewriting of logic for the jobs overlay.
> 
> I'm not opposed to this or anything, I just wanted to make sure you understood the scope that such a change would involve. I agree that it would be better long-term to only poll memory on a log line, but that's a big difference compared to the functionality this is meant to replicate.

Ok, in that case, I'd rather keep it the way it is. On top of Cactbot, there could be other downstream consumers that expect the existing behavior. On top of that, there's the annoying corner case of TEA time stop causing a slight issue for time-based gauge elements, since you can't just count down internally and still arrive at the correct result.

**Comment by valarnin** [comment]
I think I have a better way to handle the polling/logic here that would satisfy both thoughts:

1. Run a timer/thread loop that updates every 16ms (to match existing cactbot update timing)
2. On timer update:
  a. If we have no gauge data currently, fetch it, cache it along with current timestamp, and emit event
  b. Every set interval (maybe 1000ms?), update gauge data from memory regardless of paused status, emit event
  c. If we have gauge data and we're not paused, update timers on it based on last update timestamp
3. Listen for log lines
  a. On status gain or status loss, if it's the `Temporal Displacement` status effect, pause or unpause
4. Listen for combat changed event
  a. On in game combat == true, unpause
  b. On in game combat == false, pause

**Comment by xpdota** [comment]
> I'm not sure I get why `JobGaugeChanged` would be emitted every (x/tbd/1000)ms if the gauge hasn't changed. Or are you saying the event would only be emitted if the data has changed since it was last emitted? (So, basically, more of a "slow mode" than a pause?)

The definition of "changed" here is a bit ambiguous.

For timer-based gauges (e.g. WHM lilies), the game constantly updates the gauge in memory as the timer ticks up, even though the server isn't necessarily sending you more gauge packets. The server would only send you a packet (IIRC, been a bit since I tested) when you either use a lily, or when a full lily is recharged. So the value is constantly changing (unless you're capped) for a lot of jobs, despite there not being new packets. 

In other words, a packet generally corresponds to some kind of gauge state change, but the opposite is not true for many jobs. For those jobs, re-emitting every x ms is effectively the same as re-emitting when there is an in-memory change.

It's possible to make an overlay work for these jobs by using gauge log lines + an internal timer to keep track of the expected state in between packets. My understanding is that some overlays do not do this - rather, they just use a constant stream of gauge data.

**Comment by valarnin** [comment]
To clarify a bit, the goal of the proposed changes above would be to only read game memory on a few occasions:

1. When a gauge line/packet is received, which is generally whenever a non-timer part of the gauge is updated, e.g. WHM uses lily, or WHM gains lily
2. When the gauge needs to be force updated to ensure it's in sync, e.g. combat start/end, or temporal stasis gain/lose

In all other cases, instead of reading from game memory, we simulate the gauge update directly in code.

**Comment by valarnin** [comment]
Upon further review of the raw data and how it plays out network-wise, we don't even need to read memory on gauge packet received because that ***does*** include the current timer values. We can basically take the raw packet data (or reconstruct from `31` lines, trim the first byte which contains the player's job, left pad with 8 null bytes, and then directly cast/marshal that byte array to the ClientStructs struct, and that becomes our new gauge data. 

There's even a network packet that likely indicates that the gauge timers should stop due to Temporal Stasis, though I'm not sure exactly what values here correspond to stopping the gauge without a bit more research, which means we could probably simulate this entirely without reading memory:

```
4096|2024-02-14T00:44:44.1910000-05:00|ActorControl|10001234|GainEffect|0000045F|00000000|00000000|00000000|00000000|cadb116292716869
4096|2024-02-14T00:44:44.1910000-05:00|PlayerStateFlags|10001234|1000021A|00000004|00000000|00000000|3c81cf56b1a6413c
4096|2024-02-14T00:44:53.1870000-05:00|ActorControl|10001234|LoseEffect|0000045F|00000000|E0000000|00000000|00000000|756185b6fa23c88b
4096|2024-02-14T00:44:53.1870000-05:00|PlayerStateFlags|10001234|10000202|00000000|00000000|00000000|97e7787ff4bc1a45
```

**Comment by wexxlee** [comment]
> There's even a network packet that likely indicates that the gauge timers should stop due to Temporal Stasis, though I'm not sure exactly what values here correspond to stopping the gauge without a bit more research, which means we could probably simulate this entirely without reading memory:

Was just typing out a longer reply (thanks for the clarification above, btw - very helpful), but this sounds promising.

Maybe worth considering: when a timer is kicked off, OP could poll memory at some extended interval (500ms? 1000ms?) to align its simulated timer (and for that matter, other gauge values?) with the in-memory timer, just to be safe.  And could then stop polling once the timer reaches 0.  That'd be significantly less intensive than repeated 16ms polling, but would still ensure that nothing wonky happened (and would negate the need to even care about gauge-stop packets, although perhaps redundancy is a good idea).

**Comment by valarnin** [comment]
Hmm, looks like the gauge pause isn't directly controlled by those network packets after all, I'm seeing the three flags that vary set for other mechanics which bind or stun the player, some of which I've got recorded footage for showing the gauge still building...

**Comment by xpdota** [comment]
I wonder if it's possible to do it off the status effects alone. I notice that 0x45F and the identically-named 0x875 both have a 'Flags' value of 26 (https://xivapi.com/Status/1119 and https://xivapi.com/Status/2165). 0x384 does not, but I don't know if that one actually does the same thing or not. 

Now, you'd *think* something called 'flags' would be a bitmap, and one of the bits would correspond to the time freeze, but every value between 0-47 except for 11 comes up at least once. Could be a total red herring.

I'd imagine the 'PlayerStateFlags' thing is probably just an animation freeze or similar.

**Comment by wexxlee** [comment]
> I wonder if it's possible to do it off the status effects alone. I notice that 0x45F and the identically-named 0x875 both have a 'Flags' value of 26 (https://xivapi.com/Status/1119 and https://xivapi.com/Status/2165). 0x384 does not, but I don't know if that one actually does the same thing or not.

If the OP timer will be paused by status effects as opposed to a mem read, then I'd be more in favor of hardcoding the effect ids than relying on attribute data that we're not 100% sure of.  I can't think offhand of any other effect in the game that pauses gauge timers like Temporal Displacement, but if we do get another one down the road, we can always revisit (and have a second data point to confirm the theory about `Flags`).

**Comment by valarnin** [comment]
I spent some time this evening rewriting the event/update logic to work by triggering purely off of network lines, which can technically work, but it's a lot of extra logic that we'd have to add and maintain per-job. Just for WHM (which is the only job I have handled so far), the code looks like:
```cs
                case JobGaugeJob.WHM:
                    {
                        var castGauge = (WhiteMageGauge)currentJobGauge;
                        if (castGauge.Lily == 3) return;
                        castGauge.LilyTimer += (short) delta;
                        if (castGauge.LilyTimer > 20000)
                        {
                            // Additional lily gain will be sent via network packet and has already been increased by this point
                            castGauge.LilyTimer -= 20000;
                        }
                    }
                    return;
```

I think that maybe it's better to keep the memory reading logic in this specific case (Job Gauge data in general). If we're really concerned for performance here we can revisit, since it's clearly possible, but I'd rather replicate the logic that has existed in cactbot for years now than have to manually code logic for every job that has a timer involved (WHM, SCH, AST, SGE, BLM, SMN, BRD, MCH, MNK, DRG, NIN, RPR, DRK, GNB).

Maybe for those specific jobs we can poll memory, and for others we can go purely based off network lines.

I'm open to whatever everyone thinks is best here. I see three potential options at this point:

1. Replicate cactbot logic 1:1 as-is (current PR code). Maybe bump refresh rate to 16ms if needed
2. Hybrid
  a. Use cactbot logic for jobs with timers
  b. Use network data for jobs without timers (basically replace the code I included above with a memory read to get data)
3. Pure network data, only sync from memory when required

**Comment by xpdota** [comment]
> I guess potentially we could shift the burden of updating the gauge timers downstream as well but I don't really want to do that since at that point it's not a full replacement for the existing event.

True that it's not a full replacement, so that could just be a "keep it the same for now" situation. 

I feel like in the long term, it's fine to shift some of that burden downstream, because you need a bunch of job-specific logic and UIs to do anything useful with the gauge data anyway. In order to be able to recreate the gauge data, you'd need the event to contain the raw data, the timestamp (or a delta from the current time), and whether or not it is paused.

Also, I just noticed that "paused" is a bit less clear - WHM's gauge pauses out of combat, but SGE's does not, for example. So there's a bit more benefit to just always polling, rather than trying to simulate and hoping you didn't miss a corner case.

**Comment by wexxlee** [comment]
> > I guess potentially we could shift the burden of updating the gauge timers downstream as well but I don't really want to do that since at that point it's not a full replacement for the existing event.
> 
> True that it's not a full replacement, so that could just be a "keep it the same for now" situation.

Yeah, I agree with keeping it the same for now.  Perhaps in an ideal world (subjective definitions apply) it would be handled differently, but since we're already asking downstream cactbot consumers to switch their event source to OP, I'm in favor of introducing as few breaking changes as possible.

> 1. Replicate cactbot logic 1:1 as-is (current PR code). Maybe bump refresh rate to 16ms if needed

I'm leaning toward this option (and it sounds like xp is as well).  Re: polling rate, I think it's okay to leave at 50ms.  No one else has weighed in with a use case for polling harder (e.g. at the cactbot 16ms rate), so 50ms seems like a reasonable place to start unless/until there are complaints?

**Comment by xpdota** [comment]
Also, it's been a while since I've looked at the logic for handling subscriptions, but is it possible to just disable the gauge logic entirely if nothing is listening for the event(s) in question?

**Comment by valarnin** [comment]
> Also, it's been a while since I've looked at the logic for handling subscriptions, but is it possible to just disable the gauge logic entirely if nothing is listening for the event(s) in question?

https://github.com/OverlayPlugin/OverlayPlugin/blob/813dafb7ad4436e9ef4428cc3714369cf72a8123/OverlayPlugin.Core/EventSourceBase.cs#L144-L147

**Comment by valarnin** [comment]
Required some reworking (moving the thread to the event source instead of the memory manager) but it will only poll/read memory when there's a subscriber now.

**Review by wexxlee — APPROVED** [review]
Looks good to me.
